src/runner/sourceRunner.ts

```diff
@@ -49,22 +48,19 @@ const DEFAULT_SOURCE_OPTIONS: IOptions = {
   throwInfiniteLoops: true
 }
 
-let previousCode = ''
+let previousCode: {
+  files: Partial<Record<string, string>>
+  entrypointFilePath: string
+} | null = null
 let isPreviousCodeTimeoutError = false
```

Comment:
i'm wondering if we can make do without `previousCode`? it seems that it is only being used to check if the current program == previous program to derive whether we need to increase the timeout. perhaps it will be sufficient to have only a boolean flag to track if there was a previous timeout? i'm not sure if i'm missing out any other use cases for this

this may be unrelated to the refactoring in this PR but i thought i would bring this up for discussion still since this may simplify `sourceRunner`

Reply:
If I'm not wrong, we only want to increase the timeout if there was a previous timeout and the code didn't change? The second condition seems to be why we store the previous code instead of just a boolean flag.

But yes, I agree. I think that whether the timeout should be increased or not should simply be a flag in the context. The responsibility of setting this flag should be moved to the frontend. It is much easier to determine whether the code has changed or not there (and also makes changing the mechanism in the future easier). An even better alternative might be to allow specifying the timeout on the context itself, so that the frontend can decide how much time a program should be given to evaluate, with js-slang only responsible for evaluating the program as per the parameters passed in. This makes the timeout behaviour more apparent to users of js-slang.